modules/build/src/main/scala/scala/build/internal/ClassCodeWrapper.scala

```diff
@@ -0,0 +1,62 @@
+package scala.build.internal
+
+case object ClassCodeWrapper extends CodeWrapper {
+  def mainClassObject(className: Name): Name =
+    Name(className.raw ++ "_sc")
+
+  private val userCodeNestingLevel = 1
+  def apply(
+    code: String,
+    pkgName: Seq[Name],
+    indexedWrapperName: Name,
+    extraCode: String,
+    scriptPath: String
+  ) = {
+    val name             = mainClassObject(indexedWrapperName).backticked
+    val wrapperClassName = indexedWrapperName.backticked + "0"
+    val mainObjectCode =
+      AmmUtil.normalizeNewlines(s"""|object $name {
+                                    |  private var args$$opt0 = Option.empty[Array[String]]
+                                    |  def args$$set(args: Array[String]): Unit = {
+                                    |    args$$opt0 = Some(args)
+                                    |  }
+                                    |  def args$$opt: Option[Array[String]] = args$$opt0
+                                    |  def args$$: Array[String] = args$$opt.getOrElse {
+                                    |    sys.error("No arguments passed to this script")
+                                    |  }
+                                    |
+                                    |  lazy val script = new $wrapperClassName
+                                    |
+                                    |  def main(args: Array[String]): Unit = {
+                                    |    args$$set(args)
+                                    |    script.hashCode() // hashCode to clear scalac warning about pure expression in statement position
+                                    |  }
+                                    |}
+                                    |
+                                    |export $name.script as ${indexedWrapperName.backticked}
+                                    |""".stripMargin)
+
+    val packageDirective =
+      if (pkgName.isEmpty) "" else s"package ${AmmUtil.encodeScalaSourcePath(pkgName)}" + "\n"
+
+    // indentation is important in the generated code, so we don't want scalafmt to touch that
+    // format: off
+    val top = AmmUtil.normalizeNewlines(s"""
+$packageDirective
+
+
+class $wrapperClassName {
```

Comment:
You might want to mark the class as `final` here. I think it helps the JVM (maybe scalac too) optimizing things.